pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,255 @@
+	return &httpProbeRunner{}
+}
+
+func (hp *httpProbeRunner) setup(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	probeSpec := getProbeSpecFromContainer(container, probeType)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+	req, err := httpprobe.NewRequestForHTTPGetAction(probeSpec.HTTPGet, &container, status.PodIP, "probe")
+	if err != nil {
+		return err
+	}
+
+	// persist
+	hp.timeout = timeout
+	hp.httpRequest = req
+	return err
+}
+
+func (hp *httpProbeRunner) run(_ context.Context, prober prober) (probe.Result, string, error) {
+	return prober.http.Probe(hp.httpRequest, hp.timeout)
+}
+
+// grpcProbeRunner holds state and logic for running GRPC probes.
+type grpcProbeRunner struct {
+	host    string
+	service string
+	port    int
+	timeout time.Duration
+}
+
+// newGRPCProbeRunner returns grpcProbeRunner which implements probeRunner.
+func newGRPCProbeRunner() *grpcProbeRunner {
+	return &grpcProbeRunner{}
+}
+
+func (gp *grpcProbeRunner) setup(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	probeSpec := getProbeSpecFromContainer(container, probeType)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	host := status.PodIP
+	service := ""
+	if probeSpec.GRPC.Service != nil {
+		service = *probeSpec.GRPC.Service
+	}
+
+	// persist
+	gp.host = host
+	gp.service = service
+	gp.port = int(probeSpec.GRPC.Port)
+	gp.timeout = timeout
+	return nil
+}
+
+func (gp *grpcProbeRunner) run(_ context.Context, prober prober) (probe.Result, string, error) {
+	return prober.grpc.Probe(gp.host, gp.service, gp.port, gp.timeout)
+}
+
+// execProbeRunner holds state and logic for running Exec probes.
+type execProbeRunner struct {
+	containerID kubecontainer.ContainerID
+	command     []string
+	timeout     time.Duration
+}
+
+// newExecProbeRunner returns execProbeRunner which implements probeRunner.
+func newExecProbeRunner() *execProbeRunner {
+	return &execProbeRunner{}
+}
+
+func (ep *execProbeRunner) setup(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	probeSpec := getProbeSpecFromContainer(container, probeType)
+	command := kubecontainer.ExpandContainerCommandOnlyStatic(probeSpec.Exec.Command, container.Env)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	c, _ := podutil.GetContainerStatus(status.ContainerStatuses, container.Name)
+	containerID := kubecontainer.ParseContainerID(c.ContainerID)
+
+	// persist
+	ep.command = command
+	ep.timeout = timeout
+	ep.containerID = containerID
+	return nil
+}
+
+func (ep *execProbeRunner) run(ctx context.Context, prober prober) (probe.Result, string, error) {
+	return prober.exec.Probe(&execInContainer{run: func() ([]byte, error) {
+		return prober.runner.RunInContainer(ctx, ep.containerID, ep.command, ep.timeout)
+	}})
+}
+
+// emptyProbeRunner holds state and logic for running Exec probes.
+type emptyProbeRunner struct {
+	isProbeSpecNil   bool
```

Comment:
just have an error string here

Reply:
@SergeyKanzhelev I'm handling two cases here
1. Nil/Empty Probe Spec
    returning success with no error
    
2. Unknown Kind/Handle
    returning unknown with error
    
I need a boolean to determine the return type, I can also create a new `unknownProbeRunner` object then I can use an error string as part of the structure.

Reply:
@SergeyKanzhelev 
UPDATE:

Added a dummyProbeRunner which handles the following corner cases:
1. Unknown Probe Type
2. Nil Probe Spec
3. Unknown Probe Handler

I have also added test cases for these `probeRunner`